File: scripts/_template_helpers/lint_template.py
#!/usr/bin/env python3
"""Template-update lint subroutines.

Run as: lint_template.py --root <repo-root>

Emits LINT|<level>|<file>|<msg> lines. Exit 0 = clean, 1 = warnings, 2 = errors.
"""
from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from datetime import datetime, timedelta, timezone
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))
import bootstrap_hash  # noqa: E402
import manifest_parse  # noqa: E402

REQUIRED_MIG_HEADERS = {"migration", "requires", "touches", "idempotent"}
REQUIRED_PROMPT_KEYS = {"id", "requires", "scope_glob", "risk"}
RISK_VALUES = {"low", "medium", "high"}
BLOCKED_PATTERNS = ("secrets/", "themes/", ".awiki/", ".git/")


def emit(level: str, file: str, msg: str) -> None:
    print(f"LINT|{level}|{file}|{msg}")


def lint_manifest(root: Path) -> tuple[int, int]:
    errors = warnings = 0
    mp = root / "template.manifest.toml"
    if not mp.is_file():
        emit("error", "template.manifest.toml", "missing")
        return (1, 0)
    m = manifest_parse.load(mp)
    for strat, globs in m.get("strategies", {}).items():
        seen: set[str] = set()
        for g in globs:
            if g in seen:
                emit("warning", "template.manifest.toml", f"duplicate glob in {strat}: {g}")
                warnings += 1
            seen.add(g)
    return (errors, warnings)


def lint_migrations(root: Path) -> tuple[int, int]:
    errors = warnings = 0
    mig_dir = root / "migrations"
    if not mig_dir.is_dir():
        return (0, 0)
    pat = re.compile(r"^(\d{4}-[a-z0-9-]+\.(?:sh|prompt\.md)|schema-\d+-to-\d+\.sh|README\.md|\.gitkeep)$")
    for entry in mig_dir.iterdir():
        if not entry.is_file():
            continue
        if not pat.match(entry.name):
            emit("error", str(entry.relative_to(root)), "migration filename does not match pattern")
            errors += 1
            continue
        if entry.name.endswith(".sh") and not entry.name.startswith("schema-"):
            text = entry.read_text(encoding="utf-8", errors="replace")
            headers: set[str] = set()
            for line in text.splitlines():
                if not line.startswith("#"):
                    if line.strip() == "" or line.startswith("#!"):
                        continue
                    break
                m_hdr = re.match(r"#\s*([a-z_]+)\s*:", line)
                if m_hdr:
                    headers.add(m_hdr.group(1))
            missing = REQUIRED_MIG_HEADERS - headers
            if missing:
                emit("error", str(entry.relative_to(root)), f"missing header keys: {sorted(missing)}")
                errors += 1
            for line in text.splitlines():
                m_t = re.match(r"#\s*touches\s*:\s*(.*)$", line)
                if m_t:
                    for g in m_t.group(1).split():
                        for blocked in BLOCKED_PATTERNS:
                            if g.startswith(blocked) or g == blocked.rstrip("/"):
                                emit("error", str(entry.relative_to(root)),
                                     f"touches: includes blocked pattern {g}")
                                errors += 1
        elif entry.name.endswith(".prompt.md"):
            text = entry.read_text(encoding="utf-8", errors="replace")
            fm: dict[str, str] = {}
            in_fm = False
            seen_open = False
            for line in text.splitlines():
                if line.strip() == "---":
                    if not seen_open:
                        seen_open = True
                        in_fm = True
                        continue
                    else:
                        in_fm = False
                        break
                if in_fm and ":" in line:
                    k, v = line.split(":", 1)
                    fm[k.strip()] = v.strip().strip('"').strip("'")
            missing = REQUIRED_PROMPT_KEYS - fm.keys()
            if missing:
                emit("error", str(entry.relative_to(root)), f"missing frontmatter keys: {sorted(missing)}")
                errors += 1
            risk = fm.get("risk", "")
            if risk and risk not in RISK_VALUES:
                emit("error", str(entry.relative_to(root)), f"risk must be low|medium|high: {risk}")
                errors += 1
            scope = fm.get("scope_glob", "")
            for blocked in BLOCKED_PATTERNS:
                if scope.startswith(blocked) or scope == blocked.rstrip("/"):
                    emit("error", str(entry.relative_to(root)),
                         f"scope_glob points at blocked pattern: {scope}")
                    errors += 1
    return (errors, warnings)


def lint_pending_prompts(root: Path) -> tuple[int, int]:
    errors = warnings = 0
    pp = root / ".awiki" / "pending-prompts"
    if not pp.is_dir():
        return (0, 0)
    cutoff = datetime.now(timezone.utc) - timedelta(days=14)
    for p in pp.glob("*.md"):
        mtime = datetime.fromtimestamp(p.stat().st_mtime, tz=timezone.utc)
        if mtime < cutoff:
            emit("warning", str(p.relative_to(root)),
                 f"pending prompt older than 14 days (mtime={mtime.isoformat()})")
            warnings += 1
    return (errors, warnings)


def lint_provenance_drift(root: Path) -> tuple[int, int]:
    pj = root / ".awiki" / "template.json"
    if not pj.is_file():
        return (0, 0)
    d = json.loads(pj.read_text(encoding="utf-8"))
    warnings = 0
    if d.get("repo") != d.get("original_repo"):
        emit("warning", ".awiki/template.json",
             f"repo differs from original_repo (repo={d.get('repo')}, original={d.get('original_repo')})")
        warnings += 1
    pinned = d.get("commit", "")
    if pinned and not (root / ".awiki" / "template-cache" / pinned).is_dir():
        emit("warning", ".awiki/template.json",
             f"pinned commit {pinned[:12]} not in template-cache; will auto-recover on next update")
        warnings += 1
    return (0, warnings)


def lint_orphans(root: Path) -> tuple[int, int]:
    """_fetch and _scratch-merge orphans outside an active --continue session."""
    warnings = 0
    fetch_state = root / ".awiki" / "template-cache" / "_fetch" / ".update-state.json"
    if fetch_state.is_file():
        emit("warning", ".awiki/template-cache/_fetch/.update-state.json",
             "orphaned state file (in-progress update or crashed run); use --continue or --abort")
        warnings += 1
    scratch = root / ".awiki" / "template-cache" / "_fetch" / "_scratch-merge"
    if scratch.is_dir():
        emit("warning", ".awiki/template-cache/_fetch/_scratch-merge",
             "orphaned plan-time scratch dir; safe to delete")
        warnings += 1
    return (0, warnings)


def lint_pending_prompt_drift(root: Path) -> tuple[int, int]:
    """Pending-prompt mtime newer than most recent applied_migrations entry for same id."""
    pp = root / ".awiki" / "pending-prompts"
    pj = root / ".awiki" / "template.json"
    if not pp.is_dir() or not pj.is_file():
        return (0, 0)
    d = json.loads(pj.read_text(encoding="utf-8"))
    applied_ids = {m.get("id"): m for m in d.get("applied_migrations", [])}
    warnings = 0
    for p in pp.glob("*.md"):
        mid = p.stem.removesuffix(".prompt")
        if mid in applied_ids:
            mtime = datetime.fromtimestamp(p.stat().st_mtime, tz=timezone.utc)
            emit("warning", str(p.relative_to(root)),
                 f"pending prompt for {mid} present but applied_migrations[] already records it (mtime={mtime.isoformat()}) — agent may have run without recording")
            warnings += 1
    return (0, warnings)


def lint_prompt_body_scope(root: Path) -> tuple[int, int]:
    """Pending-prompt body must not reference paths outside its declared scope_glob.

    Specifically: the resolved scope listed in the prompt body must not include
    paths in BLOCKED_PATTERNS (secrets/, themes/, .awiki/, .git/).
    """
    pp = root / ".awiki" / "pending-prompts"
    if not pp.is_dir():
        return (0, 0)
    errors = 0
    for p in pp.glob("*.md"):
        text = p.read_text(encoding="utf-8")
        in_fm = False
        seen_open = False
        scope = ""
        for line in text.splitlines():
            if line.strip() == "---":
                if not seen_open:
                    seen_open = True
                    in_fm = True
                    continue
                else:
                    in_fm = False
                    break
            if in_fm and line.startswith("scope_glob:"):
                scope = line.split(":", 1)[1].strip().strip('"').strip("'")
        # Parse "## Resolved scope" body section.
        resolved: list[str] = []
        in_resolved = False
        for line in text.splitlines():
            stripped = line.strip()
            if stripped.startswith("## ") and "resolved scope" in stripped.lower():
                in_resolved = True
                continue
            if in_resolved:
                if stripped.startswith("## "):
                    in_resolved = False
                elif stripped.startswith("- "):
                    resolved.append(stripped[2:].strip())
        if scope:
            for rel in resolved:
                for blocked in BLOCKED_PATTERNS:
                    if rel.startswith(blocked):
                        emit("error", str(p.relative_to(root)),
                             f"resolved scope includes blocked path: {rel}")
                        errors += 1
                        break
    return (errors, 0)


def lint_availability(root: Path) -> tuple[int, int]:
    """Run availability.py --check; output goes through to stdout. Never fails the run."""
    subprocess.run(
        ["python3", str(HERE / "availability.py"), "--root", str(root), "--check"],
        check=False,
    )
    return (0, 0)


def main() -> int:
    p = argparse.ArgumentParser()
    p.add_argument("--root", required=True, type=Path)
    args = p.parse_args()
    total_e = total_w = 0
    for fn in (
        lint_manifest,
        lint_migrations,
        lint_pending_prompts,
        lint_provenance_drift,
        lint_orphans,
        lint_pending_prompt_drift,
        lint_prompt_body_scope,
        lint_availability,
    ):
        e, w = fn(args.root)
        total_e += e
        total_w += w
    if total_e > 0:
        return 2
    if total_w > 0:
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: internal/template/lint_template.go
package template

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

// LintLevel is the severity tag used in LINT|<level>|... output lines.
type LintLevel string

const (
	LintLevelError   LintLevel = "error"
	LintLevelWarning LintLevel = "warning"
	LintLevelInfo    LintLevel = "info"
)

// LintMessage is a single LINT|... line, captured before formatting.
type LintMessage struct {
	Level LintLevel
	File  string
	Msg   string
}

// Format returns the canonical "LINT|level|file|msg" wire format.
func (m LintMessage) Format() string {
	return fmt.Sprintf("LINT|%s|%s|%s", m.Level, m.File, m.Msg)
}

// LintReport captures every emitted message plus the cumulative
// error/warning counts. Mirrors the way `lint_template.py` aggregates
// before computing the exit code.
type LintReport struct {
	Messages []LintMessage
	Errors   int
	Warnings int
}

func (r *LintReport) emit(level LintLevel, file, msg string) {
	r.Messages = append(r.Messages, LintMessage{Level: level, File: file, Msg: msg})
	switch level {
	case LintLevelError:
		r.Errors++
	case LintLevelWarning:
		r.Warnings++
	}
}

// ExitCode returns the Python oracle exit code: 0 clean, 1 warnings,
// 2 errors.
func (r *LintReport) ExitCode() int {
	if r.Errors > 0 {
		return 2
	}
	if r.Warnings > 0 {
		return 1
	}
	return 0
}

// LintTemplate runs every check `lint_template.py` emits in order. The
// availability check is delegated to the caller via lsRemote (matching
// the same wiring used by CheckAvailability); pass nil to skip it.
//
// noCheckEnv mirrors AWIKI_NO_TEMPLATE_CHECK; pass "" to enable.
// now is the wall clock used for pending-prompt staleness.
func LintTemplate(root string, now time.Time, noCheckEnv string, lsRemote LsRemoteFunc) (*LintReport, error) {
	r := &LintReport{}
	if err := lintManifest(r, root); err != nil {
		return nil, err
	}
	if err := lintMigrations(r, root); err != nil {
		return nil, err
	}
	if err := lintPendingPrompts(r, root, now); err != nil {
		return nil, err
	}
	if err := lintProvenanceDrift(r, root); err != nil {
		return nil, err
	}
	if err := lintOrphans(r, root); err != nil {
		return nil, err
	}
	if err := lintPendingPromptDrift(r, root); err != nil {
		return nil, err
	}
	if err := lintPromptBodyScope(r, root); err != nil {
		return nil, err
	}
	if err := lintAvailability(r, root, noCheckEnv, now, lsRemote); err != nil {
		return nil, err
	}
	return r, nil
}

func lintManifest(r *LintReport, root string) error {
	mp := filepath.Join(root, "template.manifest.toml")
	if !isFile(mp) {
		r.emit(LintLevelError, "template.manifest.toml", "missing")
		return nil
	}
	m, err := LoadManifest(mp)
	if err != nil {
		return err
	}
	// Walk strategies in precedence order so warnings are deterministic.
	seenStrats := map[string]bool{}
	walk := func(strat string) {
		globs := m.Strategies[strat]
		seen := map[string]bool{}
		for _, g := range globs {
			if seen[g] {
				r.emit(LintLevelWarning, "template.manifest.toml",
					fmt.Sprintf("duplicate glob in %s: %s", strat, g))
			}
			seen[g] = true
		}
	}
	for _, s := range strategyPrecedence {
		seenStrats[s] = true
		walk(s)
	}
	var extra []string
	for s := range m.Strategies {
		if !seenStrats[s] {
			extra = append(extra, s)
		}
	}
	sort.Strings(extra)
	for _, s := range extra {
		walk(s)
	}
	return nil
}

var migrationFilenameRe = regexp.MustCompile(`^(\d{4}-[a-z0-9-]+\.(?:sh|prompt\.md)|schema-\d+-to-\d+\.sh|README\.md|\.gitkeep)$`)

func lintMigrations(r *LintReport, root string) error {
	migDir := filepath.Join(root, "migrations")
	info, err := os.Stat(migDir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	if !info.IsDir() {
		return nil
	}
	entries, err := os.ReadDir(migDir)
	if err != nil {
		return err
	}
	// Sort for deterministic emit order. Python uses iterdir which is
	// filesystem-order on POSIX; bats fixtures don't depend on order
	// but our tests will, so sort explicitly.
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)

	for _, name := range names {
		full := filepath.Join(migDir, name)
		rel := relPath(root, full)
		if !migrationFilenameRe.MatchString(name) {
			r.emit(LintLevelError, rel, "migration filename does not match pattern")
			continue
		}
		if strings.HasSuffix(name, ".sh") && !strings.HasPrefix(name, "schema-") {
			data, err := os.ReadFile(full)
			if err != nil {
				return err
			}
			text := string(data)
			headers := map[string]bool{}
			for _, line := range strings.Split(text, "\n") {
				if !strings.HasPrefix(line, "#") {
					trimmed := strings.TrimSpace(line)
					if trimmed == "" || strings.HasPrefix(line, "#!") {
						continue
					}
					break
				}
				m := migrationHeaderRe.FindStringSubmatch(line)
				if m != nil {
					headers[m[1]] = true
				}
			}
			missing := []string{}
			for _, k := range MigrationRequiredHeaderKeys {
				if !headers[k] {
					missing = append(missing, k)
				}
			}
			sort.Strings(missing)
			if len(missing) > 0 {
				r.emit(LintLevelError, rel, fmt.Sprintf("missing header keys: %s", pythonStrList(missing)))
			}
			// Touches: blocked-pattern check.
			touchesRe := regexp.MustCompile(`^#\s*touches\s*:\s*(.*)$`)
			for _, line := range strings.Split(text, "\n") {
				m := touchesRe.FindStringSubmatch(line)
				if m == nil {
					continue
				}
				for _, g := range strings.Fields(m[1]) {
					for _, blocked := range MigrationBlockedPatterns {
						if strings.HasPrefix(g, blocked) || g == strings.TrimRight(blocked, "/") {
							r.emit(LintLevelError, rel,
								fmt.Sprintf("touches: includes blocked pattern %s", g))
						}
					}
				}
			}
		} else if strings.HasSuffix(name, ".prompt.md") {
			data, err := os.ReadFile(full)
			if err != nil {
				return err
			}
			fm := MigrationParseFrontmatter(string(data))
			required := []string{"id", "requires", "scope_glob", "risk"}
			missing := []string{}
			for _, k := range required {
				if _, ok := fm[k]; !ok {
					missing = append(missing, k)
				}
			}
			sort.Strings(missing)
			if len(missing) > 0 {
				r.emit(LintLevelError, rel, fmt.Sprintf("missing frontmatter keys: %s", pythonStrList(missing)))
			}
			riskValues := map[string]bool{"low": true, "medium": true, "high": true}
			risk := fm["risk"]
			if risk != "" && !riskValues[risk] {
				r.emit(LintLevelError, rel, fmt.Sprintf("risk must be low|medium|high: %s", risk))
			}
			scope := fm["scope_glob"]
			for _, blocked := range MigrationBlockedPatterns {
				if strings.HasPrefix(scope, blocked) || scope == strings.TrimRight(blocked, "/") {
					r.emit(LintLevelError, rel, fmt.Sprintf("scope_glob points at blocked pattern: %s", scope))
				}
			}
		}
	}
	return nil
}

func lintPendingPrompts(r *LintReport, root string, now time.Time) error {
	pp := filepath.Join(root, ".awiki", "pending-prompts")
	info, err := os.Stat(pp)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	if !info.IsDir() {
		return nil
	}
	cutoff := now.UTC().Add(-14 * 24 * time.Hour)
	entries, err := os.ReadDir(pp)
	if err != nil {
		return err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)
	for _, name := range names {
		full := filepath.Join(pp, name)
		fi, err := os.Stat(full)
		if err != nil {
			return err
		}
		mtime := fi.ModTime().UTC()
		if mtime.Before(cutoff) {
			r.emit(LintLevelWarning, relPath(root, full),
				fmt.Sprintf("pending prompt older than 14 days (mtime=%s)", isoFormat(mtime)))
		}
	}
	return nil
}

func lintProvenanceDrift(r *LintReport, root string) error {
	pj := filepath.Join(root, ".awiki", "template.json")
	if !isFile(pj) {
		return nil
	}
	prov, err := LoadProvenance(pj)
	if err != nil {
		return nil // mirror Python: bare json.loads, raise would be uncaught — but treat decode errors as no-op for now.
	}
	if prov.Repo != prov.OriginalRepo {
		r.emit(LintLevelWarning, ".awiki/template.json",
			fmt.Sprintf("repo differs from original_repo (repo=%s, original=%s)", prov.Repo, prov.OriginalRepo))
	}
	pinned := prov.Commit
	if pinned != "" {
		cacheDir := filepath.Join(root, ".awiki", "template-cache", pinned)
		if !isDir(cacheDir) {
			short := pinned
			if len(short) > 12 {
				short = short[:12]
			}
			r.emit(LintLevelWarning, ".awiki/template.json",
				fmt.Sprintf("pinned commit %s not in template-cache; will auto-recover on next update", short))
		}
	}
	return nil
}

func lintOrphans(r *LintReport, root string) error {
	state := filepath.Join(root, ".awiki", "template-cache", "_fetch", ".update-state.json")
	if isFile(state) {
		r.emit(LintLevelWarning, ".awiki/template-cache/_fetch/.update-state.json",
			"orphaned state file (in-progress update or crashed run); use --continue or --abort")
	}
	scratch := filepath.Join(root, ".awiki", "template-cache", "_fetch", "_scratch-merge")
	if isDir(scratch) {
		r.emit(LintLevelWarning, ".awiki/template-cache/_fetch/_scratch-merge",
			"orphaned plan-time scratch dir; safe to delete")
	}
	return nil
}

func lintPendingPromptDrift(r *LintReport, root string) error {
	pp := filepath.Join(root, ".awiki", "pending-prompts")
	pj := filepath.Join(root, ".awiki", "template.json")
	if !isDir(pp) || !isFile(pj) {
		return nil
	}
	prov, err := LoadProvenance(pj)
	if err != nil {
		return nil
	}
	applied := map[string]bool{}
	for _, m := range prov.AppliedMigrations {
		applied[m.ID] = true
	}
	entries, err := os.ReadDir(pp)
	if err != nil {
		return err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)
	for _, name := range names {
		stem := strings.TrimSuffix(name, ".md")
		stem = strings.TrimSuffix(stem, ".prompt")
		if !applied[stem] {
			continue
		}
		full := filepath.Join(pp, name)
		fi, err := os.Stat(full)
		if err != nil {
			return err
		}
		mtime := fi.ModTime().UTC()
		r.emit(LintLevelWarning, relPath(root, full),
			fmt.Sprintf("pending prompt for %s present but applied_migrations[] already records it (mtime=%s) — agent may have run without recording",
				stem, isoFormat(mtime)))
	}
	return nil
}

func lintPromptBodyScope(r *LintReport, root string) error {
	pp := filepath.Join(root, ".awiki", "pending-prompts")
	if !isDir(pp) {
		return nil
	}
	entries, err := os.ReadDir(pp)
	if err != nil {
		return err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)
	for _, name := range names {
		full := filepath.Join(pp, name)
		data, err := os.ReadFile(full)
		if err != nil {
			return err
		}
		text := string(data)
		fm := MigrationParseFrontmatter(text)
		scope := fm["scope_glob"]
		if scope == "" {
			continue
		}
		// Parse "## Resolved scope" body section.
		var resolved []string
		inResolved := false
		for _, line := range strings.Split(text, "\n") {
			stripped := strings.TrimSpace(line)
			if strings.HasPrefix(stripped, "## ") && strings.Contains(strings.ToLower(stripped), "resolved scope") {
				inResolved = true
				continue
			}
			if inResolved {
				if strings.HasPrefix(stripped, "## ") {
					inResolved = false
				} else if strings.HasPrefix(stripped, "- ") {
					resolved = append(resolved, strings.TrimSpace(stripped[2:]))
				}
			}
		}
		for _, rel := range resolved {
			for _, blocked := range MigrationBlockedPatterns {
				if strings.HasPrefix(rel, blocked) {
					r.emit(LintLevelError, relPath(root, full),
						fmt.Sprintf("resolved scope includes blocked path: %s", rel))
					break
				}
			}
		}
	}
	return nil
}

// lintAvailability runs the availability check and forwards the LINT|
// info|template|... line into the report (without affecting the exit
// code). Network failures are silent. Pass nil lsRemote to skip
// entirely.
func lintAvailability(r *LintReport, root string, noCheckEnv string, now time.Time, lsRemote LsRemoteFunc) error {
	if lsRemote == nil {
		return nil
	}
	res, err := CheckAvailability(root, noCheckEnv, now, lsRemote)
	if err != nil {
		return err
	}
	if res.Skipped || res.Message == "" {
		return nil
	}
	// availability already returns a fully-formatted "LINT|info|..." line.
	// We convert that back into a LintMessage so the report stays homogeneous.
	parts := strings.SplitN(res.Message, "|", 4)
	if len(parts) == 4 && parts[0] == "LINT" {
		r.Messages = append(r.Messages, LintMessage{
			Level: LintLevel(parts[1]),
			File:  parts[2],
			Msg:   parts[3],
		})
	}
	return nil
}

func relPath(root, full string) string {
	rel, err := filepath.Rel(root, full)
	if err != nil {
		return full
	}
	return filepath.ToSlash(rel)
}

func isDir(p string) bool {
	fi, err := os.Stat(p)
	if err != nil {
		return false
	}
	return fi.IsDir()
}

// pythonStrList renders a sorted list the same way Python's
// `str(sorted(list))` does: ['a', 'b'] (single quotes, comma-space).
func pythonStrList(items []string) string {
	var b strings.Builder
	b.WriteByte('[')
	for i, s := range items {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteByte('\'')
		b.WriteString(s)
		b.WriteByte('\'')
	}
	b.WriteByte(']')
	return b.String()
}

// isoFormat mirrors Python `datetime.isoformat()` for an aware UTC
// instant: YYYY-MM-DDTHH:MM:SS[.ffffff]+00:00. Sub-second precision is
// only emitted if non-zero (matching CPython behaviour).
func isoFormat(t time.Time) string {
	t = t.UTC()
	micros := t.Nanosecond() / 1000
	if micros == 0 {
		return t.Format("2006-01-02T15:04:05+00:00")
	}
	return fmt.Sprintf("%s.%06d+00:00", t.Format("2006-01-02T15:04:05"), micros)
}
